Nama  : Reza Hilmi Dafa

NIM   : 24060121130044

Nama Dataset : auction verification

Link Dataset : https://archive.ics.uci.edu/dataset/713/auction+verification

Baris ini membaca dataset Iris dari google drive

In [2]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
names = ["process.b1.capacity", "process.b2.capacity", "process.b3.capacity", "process.b4.capacity", "property.price", "property.product", "property.winner", "verification.result", "verification.time"]
dataset = pd.read_csv ('/content/drive/MyDrive/ML/auction+verification.csv', names=names, header=0)

In [4]:
dataset.head(5)

process.b1.capacity  process.b2.capacity  process.b3.capacity  \
0                    0                    0                    2   
1                    0                    0                    2   
2                    0                    0                    2   
3                    0                    0                    2   
4                    0                    0                    2   

   process.b4.capacity  property.price  property.product  property.winner  \
0                    1              59                 1                0   
1                    1              59                 2                0   
2                    1              59                 4                0   
3                    1              59                 6                0   
4                    1              60                 1                0   

   verification.result  verification.time  
0                False         163.316667  
1                False         200.860000  
2                False         154.888889  
3                False         108.640000  
4                 True          85.466667

In [5]:
dataset.shape

(2043, 9)

In [6]:
dataset_numeric = dataset.astype(int)

Baris di bawah ini mengimpor modul model_selection dari scikit-learn

In [7]:
from sklearn import model_selection

-Menggunakan atribut values dari objek DataFrame Pandas (dataset), kita mengonversi dataset ke dalam bentuk array NumPy

-Memisahkan dataset menjadi dua bagian:

X: Variabel independen (fitur) yang terdiri dari kolom 0 hingga 7.

Y: Variabel dependen (kelas/target) yang terdiri dari kolom ke-7.

-kemudian menggunakan fungsi train_test_split dari scikit-learn, dataset (X dan Y) dibagi menjadi dua bagian:

X_train: Data pelatihan untuk variabel independen.

X_validation: Data validasi untuk variabel independen.

Y_train: Data pelatihan untuk variabel dependen.

Y_validation: Data validasi untuk variabel dependen.

test_size=validation_size menentukan proporsi data yang akan dialokasikan sebagai data validasi (20% dalam hal ini).

random_state=seed memastikan reproduktibilitas pembagian data dengan memberikan nilai seed pada generator angka acak.

In [8]:
array = dataset_numeric.values

# Separate features and target variable
X = array[:,0:7]
Y = array[:,7]


validation_size = 0.20
seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

Kemudian hasilnya kita tampilkan

In [9]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (1634, 7)
X_validation : (409, 7)
Y_train : (1634,)
Y_validation : (409,)


Menetapkan seed yaitu 7 pada algoritma untuk memastikan bahwa hasilnya dapat direproduksi.

Menunjukkan bahwa metrik evaluasi yang digunakan untuk mengukur performa model adalah akurasi.

In [10]:
# Test options and evaluation metric
seed = 7
scoring = 'accuracy'

* mengimpor tiga model klasifikasi yang akan dievaluasi: SVC (Support Vector Machine), KNeighborsClassifier (K-Nearest Neighbors), dan GaussianNB (Naive Bayes).

* kemudian membuat list models yang berisi tuple-tuple model dan nama-nama model yang akan dievaluasi. Setiap tuple berisi nama model dan objek model tersebut.

selanjutnya

* Menggunakan loop for untuk iterasi melalui setiap model dalam list models.

* Membuat objek KFold untuk validasi silang dengan 10 lipatan (folds), menggunakan nilai seed yang telah ditentukan (random_state=seed) dan dengan pengacakan data (shuffle=True).

* Menggunakan cross_val_score untuk mengevaluasi setiap model menggunakan validasi silang. Hasil evaluasi disimpan dalam cv_results.

* Menambahkan hasil evaluasi dan nama model ke dalam list results dan names.

* Mencetak hasil evaluasi model, termasuk rata-rata (mean()) dan deviasi standar (std()).

Jadi, code di bawah ini membandingkan performa tiga model klasifikasi yang berbeda menggunakan metode validasi silang. Hasil evaluasi ini dapat memberikan wawasan tentang model mana yang paling sesuai dengan data pelatihan yang diberikan.

In [11]:
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB


models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))


results = []
names = []
for name, model in models:
  kfold = model_selection.KFold(n_splits=10, shuffle=True, random_state=seed)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)


KNN: 0.974304 (0.011502)
NB: 0.826205 (0.036847)
SVM: 0.869015 (0.032114)


mengimpor beberapa metrik evaluasi dari scikit-learn yang akan digunakan untuk mengevaluasi performa model klasifikasi.

In [20]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

Membuat objek model Support Vector Machine (svm) dari kelas SVC.

Melatih model menggunakan data pelatihan (X_train dan Y_train) menggunakan metode fit().

Menggunakan model SVM yang telah dilatih untuk membuat prediksi pada data validasi

Mencetak akurasi model dengan membandingkan prediksi dengan nilai sebenarnya menggunakan accuracy_score.

Mencetak matriks kebingungan (confusion matrix) menggunakan confusion_matrix, yang memberikan gambaran tentang seberapa baik model mengklasifikasikan kelas-kelas.

Mencetak laporan klasifikasi (classification_report), yang menyediakan informasi seperti presisi, recall, dan f1-score untuk setiap kelas.

In [22]:
knn = KNeighborsClassifier()
knn.fit(X_train, Y_train)
predictions_knn = knn.predict(X_validation)

print(accuracy_score(Y_validation, predictions_knn))
print(confusion_matrix(Y_validation, predictions_knn))
print(classification_report(Y_validation, predictions_knn))


0.9779951100244498
[[361   0]
 [  9  39]]
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       361
           1       1.00      0.81      0.90        48

    accuracy                           0.98       409
   macro avg       0.99      0.91      0.94       409
weighted avg       0.98      0.98      0.98       409



Penjelasan informasi atau kesimpulan dari data yang ditampilkan

Dari hasil evaluasi model pada data validasi, terlihat bahwa model KNN memiliki tingkat akurasi yang tinggi, mencapai 97.8%. Ini berarti model dengan metode KNN berhasil mengklasifikasikan sebagian besar data dengan benar.

True Positive (TP): 39

True Negative (TN): 361

False Positive (FP): 0

False Negative (FN): 9

Dari sini, kita dapat melihat bahwa model KNN memiliki kemampuan yang baik dalam mengidentifikasi kelas positif (1), ditunjukkan oleh nilai True Positive yang tinggi (39) dan False Negative yang rendah (9).

Precision: Kemampuan model untuk tidak memberikan label positif pada sampel yang seharusnya negatif (0.98 untuk kelas 0, 1.00 untuk kelas 1).

Recall: Kemampuan model untuk menemukan semua sampel positif (1.00 untuk kelas 0, 0.81 untuk kelas 1).

F1-score: Harmonic mean antara precision dan recall (0.99 untuk kelas 0, 0.90 untuk kelas 1).

Dapat disimpulkan bahwa model KNN memiliki performa yang sangat baik dalam mengklasifikasikan data pada dataset validasi.